lib/astutils.cpp

```diff
@@ -133,6 +133,20 @@ const Token * astIsVariableComparison(const Token *tok, const std::string &comp,
     return ret;
 }
 
+const Token * nextAfterAstRightmostLeaf(const Token * tok)
+{
+    const Token * rightmostLeaf = tok;
+    if (!rightmostLeaf || !rightmostLeaf->astOperand1())
+        return nullptr;
+    do {
+        if (rightmostLeaf->astOperand2())
+            rightmostLeaf = rightmostLeaf->astOperand2();
+        else
+            rightmostLeaf = rightmostLeaf->astOperand1();
```

Comment:
I dont think it matters for this case since `isSameExpression` is always false for increment/decrement operators.

This code was moved from valueflow.cpp to astutils.cpp.

Reply:
> This code was moved from valueflow.cpp to astutils.cpp.

I am not sure why we duplicated this functionality before.

How about breaking out the code from `Token::expressionString` that locates the start and end tokens into a new function `Token::findExpressionStartEndTokens`?

if the expression is `a[2];` I assume you want to return the `;` token? That would be handled better by `Token::findExpressionStartEndTokens` as far as I see.

Reply:
The function `Token::expressionString` seems to work differently. It will consume a larger expression that `isSameExpression` wont check for, which will lead to FPs.

Reply:
I do not understand but I have faith in you.

Reply:
I dont have an example. Its just that `Token::expressionString` uses `link()` and `goToRightParenthesis` to get to the right most token. However, this is not how `isSameExpression` works, it traverses the ast operators. 

Put another way, if `nextAfterAstRightmostLeaf` gives me a token which is a `;`, then I know I can use `isSameExpression` to check the entire statement. However, if `Token::expressionString` gives me a token which is a `;`, it may not be reachable through the ast operators, which means `isSameExpression` will fail to check the entire expression.

It seems it would be better for `Token::expressionString` to use `nextAfterAstRightmostLeaf` and then it can call `link()` or `goToRightParenthesis` on top of that.

Reply:
well I will look into it some more